host/traceh.c

```diff
@@ -314,6 +314,23 @@ oe_result_t oe_log(oe_log_level_t level, const char* fmt, ...)
     return result;
 }
 
+void* _oe_log_context = NULL;
+oe_log_callback _oe_log_callback = NULL;
+
+oe_result_t oe_set_log_callback(void* context, oe_log_callback callback)
+{
+    if (oe_mutex_lock(&_log_lock) == OE_OK)
+    {
+        _oe_log_context = context;
+        _oe_log_callback = callback;
+        oe_mutex_unlock(&_log_lock);
```

Comment:
There is a subtle thread-safety issue here. The compiler or the hardware could reorder the instructions to:
```c
_oe_log_callback = callback;
_oe_log_context = context;
```
This means that when another thread checks
```c
if (_oe_log_callback)
   ...
```

`_oe_log_context` may be NULL or might be the incorrect one.

Reply:
Can the mutex lock work as a barrier?
https://github.com/openenclave/openenclave/blob/2615de41ac44e47c0f683017135ebeccccb7fd23/host/traceh.c#L322

Reply:
It is a barrier; but the instructions between `oe_mutex_lock` and `oe_mutex_unlock` could be reordered arbitrarily.
Normally this doesn't matter; but elsewhere we are testing `_oe_log_callback` outside a lock.

Reply:
It would be good to define what the contract is. Is the user allowed to change the logging callback or set it only once?

Reply:
Yes you are right, I didn't realize the the if check is not safe. Now it is inside a lock. Thanks!